### PR TITLE
[FLINK-19852][task] Reuse TempBarrier memory between iterations

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

Reuse allocated memory segments between the iterations of an iterative job.

## Brief change log

  - fix checkstyle and IDE warnings
  - introduce a method to close `TempBarrier` without releasing memory
  - add `TempBarrier` constructor argument to use preallocated memory
  - use new method/argument in `BatchTask.resetAllInputs`

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/BatchTask.java [hunk 1]
```diff
@@ -893,21 +896,34 @@ protected void resetAllInputs() throws Exception {
...
-					if (this.inputIsAsyncMaterialized[i]) {
-						final int pages = this.materializationMemory[i];
-						@SuppressWarnings({ "unchecked", "rawtypes" })
-						TempBarrier<?> barrier = new TempBarrier(this, getInput(i), this.inputSerializers[i], memMan, ioMan, pages);
-						barrier.startReading();
-						this.tempBarriers[i] = barrier;
-						this.inputs[i] = null;
+						if (this.inputIsAsyncMaterialized[i]) {
+							final int pages = this.materializationMemory[i];
+							if (pages > allocated.size()) {
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/BatchTask.java:908]: This condition seems the wrong way around.

**Review by StephanEwen — COMMENTED** [review]
There might be one minor big in there, about a flipped comparison.

As far as I remember, the memory never changed between iterations, so to avoid accidental bugs like that, we could also have a check that the number of memory segments must be the same always.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/BatchTask.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/operators/BatchTask.java]: You are right, thanks!

I changed it to just
```Preconditions.checkState(allocated.size() <= pages);```

(`pages` should never change as you pointed out, but some of the allocated segments might have been consumed, so it's `<=` and not `==`)

**Comment by StephanEwen** [comment]
Looks good to me, +1 to merge this

**Comment by AHeise** [comment]
Also LGTM. Merging.

**Review by AHeise — APPROVED** [review]

**Pull request merged**